web/src/message_store.ts

```diff
@@ -0,0 +1,266 @@
+        return undefined;
+    }
+
+    if (typeof message_id !== "number") {
+        blueslip.error("message_store got non-number", {message_id});
+
+        // Try to soldier on, assuming the caller treats message
+        // ids as strings.
+        message_id = Number.parseFloat(message_id);
+    }
+
+    return stored_messages.get(message_id);
+}
+
+export function get_pm_emails(message: Message & {reply_to?: string; url?: string}): string {
+    const user_ids = people.pm_with_user_ids(message);
+    assert(user_ids !== undefined, "user_ids is undefined");
+    const emails = user_ids
+        .map((user_id) => {
+            const person = people.maybe_get_user_by_id(user_id);
+            if (!person) {
+                blueslip.error("Unknown user id", {user_id});
+                return "?";
+            }
+            return person.email;
+        })
+        .sort();
+
+    return emails.join(", ");
+}
+
+export function get_pm_full_names(message: Message & {reply_to?: string; url?: string}): string {
+    const user_ids = people.pm_with_user_ids(message);
+    assert(user_ids !== undefined, "user_ids is undefined");
+    const names = people.get_display_full_names(user_ids).sort();
+
+    return names.join(", ");
+}
+
+export function set_message_booleans(raw_message: RawMessage): MessageWithBooleans {
+    const {flags, ...raw_message_without_flags} = raw_message;
+
+    function convert_flag(flag_name: string): boolean {
+        return flags?.includes(flag_name) || false;
+    }
+
+    const message: MessageWithBooleans = {
+        ...raw_message_without_flags,
+        unread: !convert_flag("read"),
+        historical: convert_flag("historical"),
+        starred: convert_flag("starred"),
+        mentioned: convert_flag("mentioned") || convert_flag("wildcard_mentioned"),
+        mentioned_me_directly: convert_flag("mentioned"),
+        wildcard_mentioned: convert_flag("wildcard_mentioned"),
+        collapsed: convert_flag("collapsed"),
+        alerted: convert_flag("has_alert_word"),
+    };
+
+    // Once we have set boolean flags here, the `flags` attribute is
+    // just a distraction, so we delete it.  (All the downstream code
+    // uses booleans.)
+    return message;
+}
+
+export function update_booleans(message: MessageWithBooleans, flags: string[]): void {
+    // When we get server flags for local echo or message edits,
+    // we are vulnerable to race conditions, so only update flags
+    // that are driven by message content.
+    function convert_flag(flag_name: string): boolean {
+        return flags.includes(flag_name);
+    }
+
+    message.mentioned = convert_flag("mentioned") || convert_flag("wildcard_mentioned");
+    message.mentioned_me_directly = convert_flag("mentioned");
+    message.wildcard_mentioned = convert_flag("wildcard_mentioned");
+    message.alerted = convert_flag("has_alert_word");
+}
+
+export function update_property({property, value, info}: UpdateProperty): void {
```

Comment:
`info` is already an object wrapping named arguments for this function; we don’t need two levels of these wrappers.

This function is just a `case` statement and every caller hits a specific known branch, so it would be simpler to split this into four functions `update_sender_name`, `update_small_avatar_url`, `update_stream_name`, `update_status_emoji_info`. Then we don’t need a complicated type that combines them into one interface.

Reply:
thank you for the guidance! I will modify correspondingly